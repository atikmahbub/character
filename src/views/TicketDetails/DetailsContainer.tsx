import { Box, Stack, styled, Typography } from "@mui/material";

const DoubleColumn = styled(Box)({
  display: "flex",
  alignItems: "center",
  gap: "4px",
});

type Props = {
  data: any;
};

const DetailsContainer = ({ data }: Props) => {
  return (
    <Stack spacing={1}>
      <Typography fontWeight={700} variant="h4">
        {data?.name}
      </Typography>
      <DoubleColumn>
        <Typography fontWeight={700} variant="subtitle2">
          Status:
        </Typography>
        <Typography variant="subtitle2">{data?.status}</Typography>
      </DoubleColumn>
      <DoubleColumn>
        <Typography fontWeight={700} variant="subtitle2">
          Species:
        </Typography>
        <Typography variant="subtitle2">{data?.species}</Typography>
      </DoubleColumn>
      <DoubleColumn>
        <Typography fontWeight={700} variant="subtitle2">
          Gender:
        </Typography>
        <Typography variant="subtitle2">{data?.gender}</Typography>
      </DoubleColumn>
      <DoubleColumn>
        <Typography fontWeight={700} variant="subtitle2">
          Last Known Location:
        </Typography>
        <Typography variant="subtitle2">{data?.location.name}</Typography>
      </DoubleColumn>
      <DoubleColumn>
        <Typography fontWeight={700} variant="subtitle2">
          Type:
        </Typography>
        <Typography variant="subtitle2">
          {data?.type ? data.type : "N/A"}
        </Typography>
      </DoubleColumn>
      <Stack>
        <Typography fontWeight={700} variant="subtitle2">
          Last Episode's:
        </Typography>
        <Stack pl={5} mt={0.5}>
          {data?.episode
            .slice(0)
            .reverse()
            // eslint-disable-next-line array-callback-return
            .map((item: string, i: number) => {
              if (i < 5) {
                return (
                  <a
                    style={{ textDecoration: "none" }}
                    href={item}
                    rel="noreferrer"
                    target="_blank"
                  >
                    <Typography
                      color="indigo"
                      fontWeight={600}
                      variant="subtitle2"
                    >
                      {i + 1}: {item}
                    </Typography>
                  </a>
                );
              }
            })}
        </Stack>
      </Stack>
    </Stack>
  );
};

export default DetailsContainer;
