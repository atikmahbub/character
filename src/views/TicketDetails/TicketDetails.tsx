import { Box, Stack, styled, Typography } from "@mui/material";
import { useEffect, useReducer, Fragment } from "react";
import { useParams } from "react-router-dom";
import AddTitle from "../../components/AddTitle";
import { API } from "../../config/Api";
import {
  characterReducer,
  initialState,
  GET_CHARACTERS_PENDING,
  GET_CHARACTERS_SUCCESS,
  GET_CHARACTERS_ERROR,
} from "../../reducers/characterReducer";

const StyledImage = styled("img")({
  borderRadius: 5,
});

const DoubleColumn = styled(Box)({
  display: "flex",
  alignItems: "center",
  gap: "4px",
});

const TicketDetails = () => {
  const { id } = useParams();
  const [state, dispatch] = useReducer(characterReducer, initialState);

  useEffect(() => {
    dispatch({ type: GET_CHARACTERS_PENDING });
    if (id) {
      API.get(`${id}`)
        .then((response) => {
          dispatch({ type: GET_CHARACTERS_SUCCESS, payload: response.data });
        })
        .catch((error) => {
          dispatch({ type: GET_CHARACTERS_ERROR, payload: error });
        });
    }
  }, [id]);

  console.log(state.data);

  return (
    <Fragment>
      <AddTitle title="Character Details" backButton />
      <Box display="flex" gap={2} alignItems="flex-start" mt={10}>
        <StyledImage src={state.data?.image} alt="pic" width={350} />
        <Stack spacing={1}>
          <Typography fontWeight={700} variant="h4">
            {state.data?.name}
          </Typography>
          <DoubleColumn>
            <Typography fontWeight={700} variant="caption">
              Status:
            </Typography>
            <Typography variant="caption">{state.data?.status}</Typography>
          </DoubleColumn>
          <DoubleColumn>
            <Typography fontWeight={700} variant="caption">
              Species:
            </Typography>
            <Typography variant="caption">{state.data?.species}</Typography>
          </DoubleColumn>
          <DoubleColumn>
            <Typography fontWeight={700} variant="caption">
              Gender:
            </Typography>
            <Typography variant="caption">{state.data?.gender}</Typography>
          </DoubleColumn>
          <DoubleColumn>
            <Typography fontWeight={700} variant="caption">
              Last Known Location:
            </Typography>
            <Typography variant="caption">
              {state.data?.location.name}
            </Typography>
          </DoubleColumn>
        </Stack>
      </Box>
    </Fragment>
  );
};

export default TicketDetails;
