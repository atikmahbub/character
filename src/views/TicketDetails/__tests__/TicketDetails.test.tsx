import React from "react";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import TicketDetails from "../TicketDetails";
import * as ReactDOM from "react-dom";

const jsx = (
  <BrowserRouter>
    <TicketDetails />
  </BrowserRouter>
);

test("Render Without Crashing", () => {
  const tree = renderer.create(jsx).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Ticket Details Tests", () => {
  let container: HTMLDivElement;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
    // eslint-disable-next-line testing-library/no-render-in-setup
    ReactDOM.render(
      <BrowserRouter>
        <TicketDetails />
      </BrowserRouter>,
      container
    );
  });

  afterEach(() => {
    document.body.removeChild(container);
    container.remove();
  });

  it("Renders correctly initial document", () => {
    const title = container.querySelectorAll("h3");
    expect(title).toHaveLength(1);
    expect(title[0]).toHaveTextContent("Character Details");
  });

  it("Image Test", () => {
    const image = container.querySelectorAll("img");
    expect(image).toBeTruthy();
  });
});
