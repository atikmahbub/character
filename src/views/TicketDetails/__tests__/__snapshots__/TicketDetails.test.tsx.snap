// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Without Crashing 1`] = `
Array [
  <h3
    className="MuiTypography-root MuiTypography-h3 css-47suzj-MuiTypography-root"
  >
    Character Details
  </h3>,
  <div
    className="MuiBox-root css-qpry38"
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ArrowBackIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
      />
    </svg>
    <span
      className="MuiTypography-root MuiTypography-caption css-d1fgqh-MuiTypography-root"
    >
      Go Back
    </span>
  </div>,
  <div
    className="MuiBox-root css-tbucl4"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-wave css-1slz293-MuiSkeleton-root"
      style={
        Object {
          "height": 300,
          "width": 350,
        }
      }
    />
    <div
      className="css-1okj3ks-MuiStack-root"
    >
      <h4
        className="MuiTypography-root MuiTypography-h4 css-1313egv-MuiTypography-root"
      />
      <div
        className="MuiBox-root css-uklcv"
      >
        <span
          className="MuiTypography-root MuiTypography-caption css-sfqsx-MuiTypography-root"
        >
          Status:
        </span>
        <span
          className="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
        />
      </div>
      <div
        className="MuiBox-root css-uklcv"
      >
        <span
          className="MuiTypography-root MuiTypography-caption css-sfqsx-MuiTypography-root"
        >
          Species:
        </span>
        <span
          className="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
        />
      </div>
      <div
        className="MuiBox-root css-uklcv"
      >
        <span
          className="MuiTypography-root MuiTypography-caption css-sfqsx-MuiTypography-root"
        >
          Gender:
        </span>
        <span
          className="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
        />
      </div>
      <div
        className="MuiBox-root css-uklcv"
      >
        <span
          className="MuiTypography-root MuiTypography-caption css-sfqsx-MuiTypography-root"
        >
          Last Known Location:
        </span>
        <span
          className="MuiTypography-root MuiTypography-caption css-1sn4lm3-MuiTypography-root"
        />
      </div>
    </div>
  </div>,
]
`;
